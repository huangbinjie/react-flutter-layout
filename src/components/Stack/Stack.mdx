---
name: Stack
menu: components
---

import { Playground, Props } from 'docz'
import Container from '../Container'
import Stack from '../Stack'
import Positioned from '../Positioned'

## Stack

提供可以重叠的布局效果。子组件有 none-positioned 的组件时，高宽是 none-positioned 子组件高宽。
子组件没有 none-positioned 时，宽度会占满父容器。

## Usage

<Playground>
  <Container width="200px" height="200px" color="#ececec">
    <Stack>
      <Positioned top="0px" left="0px">左上</Positioned>
      <Positioned top="0px" right="0px">右上</Positioned>
    </Stack>
  </Container>
</Playground>

### 有 none-positioned 组件时

<Playground>
  <Container width="200px" height="200px" color="#ececec">
    <Stack>
      <Container width="100px" height="100px" color="coral">none-positioned</Container>
      <Positioned top="0px" left="0px">左上</Positioned>
      <Positioned top="0px" right="0px">右上</Positioned>
    </Stack>
  </Container>
</Playground>

## Props

<Props of={Stack} />
