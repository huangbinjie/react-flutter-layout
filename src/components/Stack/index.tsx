import React, { CSSProperties, ReactNode } from 'react'
import { style as jss } from 'typestyle'
import Positioned from '../Positioned'

export type StackProps = {
  className?: string
  style?: CSSProperties
  children?: ReactNode
}

/**
 * https://api.flutter.dev/flutter/rendering/RenderStack-class.html
 * 
 * The stack is then sized to enclose all of the non-positioned children.
 * If there are no non-positioned children, the stack becomes as large as possible.
 */
export default function Stack({ className = '', children, ...restProps }: StackProps) {
  let hasNonePositioned = false

  React.Children.forEach(children, (child: any) => {
    if (child && child.type !== Positioned) {
      hasNonePositioned = true
    }
  })

  const width = hasNonePositioned ? 'auto' : '100%'
  const height = hasNonePositioned ? 'auto' : '100%'

  const classname = jss({
    position: 'relative',
    width,
    height
  })

  return <div className={className + ' ' + classname} debug-label="Stack" {...restProps}>{children}</div>
}
