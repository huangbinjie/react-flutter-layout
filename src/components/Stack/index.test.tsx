import React from 'react'
import renderer from 'react-test-renderer'

import Stack from '.'

describe('Stack', () => {
  test('basic', () => {
    const tree = renderer.create(
      <Stack>empty content</Stack>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
