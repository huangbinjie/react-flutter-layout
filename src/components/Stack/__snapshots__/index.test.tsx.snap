// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack basic 1`] = `
<div
  className="fqub0h7"
  debug-label="Stack"
>
  empty content
</div>
`;
